model/settings/service_test.go

```diff
@@ -3,10 +3,128 @@ package settings
 import (
 	"testing"
 
+	"github.com/cozy/cozy-stack/model/instance"
+	"github.com/cozy/cozy-stack/model/token"
+	"github.com/cozy/cozy-stack/pkg/couchdb"
+	"github.com/cozy/cozy-stack/pkg/emailer"
 	"github.com/stretchr/testify/assert"
 )
 
 func TestServiceImplems(t *testing.T) {
 	assert.Implements(t, (*Service)(nil), new(SettingsService))
 	assert.Implements(t, (*Service)(nil), new(Mock))
 }
+
+func Test_StartEmailUpdate_success(t *testing.T) {
+	emailerSvc := emailer.NewMock(t)
+	instSvc := instance.NewMock(t)
+	tokenSvc := token.NewMock(t)
+	storage := newStorageMock(t)
+
+	svc := NewService(emailerSvc, instSvc, tokenSvc, storage)
+
+	inst := instance.Instance{
+		Domain: "foo.mycozy.cloud",
+	}
+
+	cmd := &UpdateEmailCmd{
+		Passphrase: []byte("some-pass"),
+		Email:      "some@email.com",
+	}
+
+	instSvc.On("CheckPassphrase", &inst, cmd.Passphrase).Return(nil).Once()
+
+	storage.On("getInstanceSettings", &inst).Return(&couchdb.JSONDoc{
+		M: map[string]interface{}{"public_name": "Jane Doe"},
```

Comment:
It is an example of what I try to explain sooner. Here, the mock is not exactly the same thing as the real implementation. It doesn't set the doctype, the ID and the revision. And it doesn't check them on `setInstanceSettings`.

In `func (s *CouchdbStorage) getInstanceSettings(inst prefixer.Prefixer) (*couchdb.JSONDoc, error) {`, if we had forgotten the line `doc.Type = consts.Settings`, the test would be green but the code wouldn't work.

Reply:
Yep, mocks doesn't works well when with unstructured types like `couchdb.JSONDoc`. That's why I tend to works with structured types and in this kind of case I would take a structured struct as argument and handle all the unstructured logic inside `setInstanceSettings`. 

Something like:
```go
type Settings  struct {
  PublicName string 
  PendingEmail string 
}

func (s *Storage) setInstanceSettings(inst prefixer.Prefixer, settings *Settings) error {
  doc := couchdb.JSONDoc{
    M: map[string]interface{}{
      "pending_email": settings.PendingEmail,
      "public_name": settings.PublicName,
    }
  }

  return couchdb.UpdateDoc(inst, doc)}
}

This ensure a strong contract between the parts.

Reply:
This is why I was talking about [in this previous comment](https://github.com/cozy/cozy-stack/pull/3988#discussion_r1231365898)

Reply:
> mocks doesn't works well when with unstructured types like couchdb.JSONDoc

Structured types help, but `couchdb.JSONDoc` has an explicit field `Type` for the doctype, and it is not filled here.

Reply:
The field `Type` is filled by the method `CouchdbStorage.getInstanceSettings` and used by `CouchdbStorage.setInstanceSettings. This field is part of the couchdb logic and only the `CouchdbStorage` should interact with it. The Service should't even be aware of this kind of internal fields. 

The fact we use the `couchdb.JSONDoc` is clearly a temporary state for me in order to make to gain some time.